File: C++/ACM/src/Buttons.cpp
#include <cstdio>

int primes[1229];
void sieve()
{
    bool isPrime[10001];
    for(int i=0; i<=10000; i++)
        isPrime[i]=true;
    for(int i=2; i*i<=10000; i++)
    {
        if(isPrime[i])
        {
            for(int j=i*i; j<=10000; j+=i)
                isPrime[j]=false;
        }
    }
    int c=0;
    for(int i=2; i<=10000; i++)
        if(isPrime[i])
            primes[c++]=i;
}
int main()
{
    sieve();
    int k;
    scanf("%d",&k);
    if(k%4==0&&k%3!=0)
    {
        printf("%d",3);
        return 0;
    }
    if((k&1)==0)
        k>>=1;
    for(int i=1; i<1229; i++)
        if(k%primes[i]==0)
        {
            printf("%d",primes[i]-1);
            return 0;
        }
    printf("%d",k-1);
    return 0;
}

File: Java/ACM contests/src/timus/Buttons.java
package timus;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Buttons {

	static int[] primes=new int[1229];
	static void sieve()
	{
		boolean []isPrime=new boolean[10001];
		Arrays.fill(isPrime, true);
		for(int i=2;i*i<=10000;i++)
		{
			if(isPrime[i])
			{
				for(int j=i*i;j<=10000;j+=i)
					isPrime[j]=false;
			}
		}
		int c=0;
		for(int i=2;i<=10000;i++)
			if(isPrime[i])
				primes[c++]=i;
	}
	public static void main(String[] args) throws NumberFormatException, IOException {
		sieve();
		BufferedReader st=new BufferedReader(new InputStreamReader(System.in));
		int k=Integer.parseInt(st.readLine());
		if(k%4==0&&k%3!=0)
		{
			System.out.println(3);
			return;
		}
		if(k%2==0)
			k/=2;
		boolean found=false;
		for(int i=1;i<primes.length;i++)
			if(k%primes[i]==0)
			{
				System.out.println(primes[i]-1);
				found=true;
				break;
			}
		if(!found)
			System.out.println(k-1);
	}
}
